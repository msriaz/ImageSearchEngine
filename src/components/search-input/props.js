import PropTypes from 'prop-types';

export const propTypes = {
  onChangeText: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  title: PropTypes.string,
};

export const defaultProps = {
  placeholder: 'search images',
  title: 'Search',
};
