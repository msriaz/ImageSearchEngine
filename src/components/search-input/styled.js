import {Button} from 'react-native';
import styled from 'styled-components/native';
import {Spinner} from '../spinner';

export const StyledTextInput = styled.TextInput`
  border: 1px;
  border-color: ${({theme}) => theme.color.lightGray};
  width: ${({theme}) => theme.windowWidth * 0.75}px;
  padding: ${({theme}) => theme.spacing.small}px;
`;

export const StyledButton = styled(Button)``;

export const StyledWrapper = styled.View`
  flex-direction: row;
  justify-content: space-between;
  margin: ${({theme}) => theme.spacing.small}px;
`;

export const StyledSpinner = styled(Spinner)`
justify-content: center;
align-self: center;

`;
export const SpinnerWrapper = styled.View`
 justify-content: center;
align-self: center;
align-items: center;
background-color: aliceblue;
`;