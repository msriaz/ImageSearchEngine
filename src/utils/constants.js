import {color as colors} from '../theme';
import {generateColorPalleteConstants} from './index';

export const APP_COLORS = generateColorPalleteConstants(colors);

export const STORAGE_KEYS = {
  HISTORY: 'history',
};
