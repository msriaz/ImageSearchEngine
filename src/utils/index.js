import {useEffect, useRef} from 'react';
import snakeCase from 'lodash/snakeCase';
/**
 *
 * @param {*} value
 * @returns previous state
 */
export const usePrevious = value => {
  const ref = useRef();
  useEffect(() => {
    ref.current = value;
  });
  return ref.current;
};

/**
 *
 * @param {*} imgData
 * @returns image url to render
 */
export const getImageUrl = imgData => {
  const {farm, server, id, secret} = imgData;
  return `https://farm${farm}.static.flickr.com/${server}/${id}_${secret}.jpg`;
};

/**
 * generate color consts from color pallette
 * @param {*} colors
 */
export const generateColorPalleteConstants = colors => {
  const colorsList = {};
  Object.keys(colors).forEach(colorKey => {
    colorsList[snakeCase(colorKey).toUpperCase()] = colors[colorKey];
  });
  return colorsList;
};

/**
 *
 * @param {*} funcName
 * @param {*} error
 */
export const reportError = (funcName, error) => {
  console.log(`Error in ${funcName}`, error);
};
