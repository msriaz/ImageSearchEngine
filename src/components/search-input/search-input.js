import React, {useEffect, useState} from 'react';
import {
  StyledTextInput,
  StyledButton,
  StyledWrapper,
  StyledSpinner,
  SpinnerWrapper,
} from './styled';
import {props, defaultProps} from './props';
import {reportError, usePrevious} from '../../utils';
import {save, load} from '../../utils/storage';
import {APP_COLORS, STORAGE_KEYS} from '../../utils/constants';
import {Spinner} from '../spinner';
/**
 * Search Input Component
 * @param {onPressSearchButton} defines the action to happen on search button press
 * @param {placeholder} defines size of the placeholder of Text Input
 * @param {rest} defines additional props to be passed to TextInput
 * @param {title} defines title of search button
 */

export const SearchInput = ({
  onPressSearchButton,
  placeholder,
  title,
  value,
  isLoading,
  rest,
}) => {
  const [text, onChangeText] = useState('');
  const previousText = usePrevious({text});

  useEffect(() => {
    onChangeText(value);
  }, [value]);

  const onPressSearchButtonHandler = () => {
    if (text !== previousText.text) {
      onPressSearchButton(text);
      saveToStorage();
    }
  };

  const saveToStorage = async () => {
    try {
      const savedHistory = await load(STORAGE_KEYS.HISTORY);
      console.log('@@@ saved history', savedHistory);
      if (savedHistory?.indexOf(text) === -1 || !savedHistory) {
        const updatedHistory = savedHistory ? [...savedHistory, text] : [text];
        await save(STORAGE_KEYS.HISTORY, updatedHistory);
      }
    } catch (error) {
      reportError('saveToStorage', error);
    }
  };

  return (
    <StyledWrapper>
      <StyledTextInput
        onChangeText={onChangeText}
        value={text}
        placeholder={placeholder}
        {...rest}
      />
      {isLoading ? (
        <SpinnerWrapper>
          <StyledSpinner
            variant={Spinner.VARIANTS.SMALL}
            color={APP_COLORS.BLUE}
          />
        </SpinnerWrapper>
      ) : (
        <StyledButton
          onPress={onPressSearchButtonHandler}
          title={title}
          disabled={text === previousText?.text}
          testID="searchBtn"
        />
      )}
    </StyledWrapper>
  );
};

SearchInput.props = props;
SearchInput.defaultProps = defaultProps;
