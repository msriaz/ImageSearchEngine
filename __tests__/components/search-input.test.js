import React from 'react';
import {fireEvent} from '@testing-library/react-native';
import {SearchInput} from '../../src/components/search-input';
import {render} from '../utils/theme-wrapper';

const onPressMock = jest.fn();

describe('SearchInput Component', () => {
  it('should match the Search Input snapshot', () => {
    const compTree = render(
      <SearchInput
        onChangeText={onPressMock}
        value="cat"
        title="search"
        placeholder="Type something"
        isLoading
        onPressSearchButton={onPressMock}
      />,
    );
    expect(compTree.toJSON()).toMatchSnapshot();
  });

  it('SearchInput should be editable', () => {
    const {getByPlaceholderText, getByTestId} = render(
      <SearchInput
        value="cat"
        title="search"
        placeholder="Type something"
        isLoading={false}
        onPressSearchButton={onPressMock}
      />,
    );
    fireEvent(getByPlaceholderText('Type something'), 'onChangeText', 'cat');
    fireEvent(getByTestId('searchBtn'), 'onPress');
    expect(onPressMock).toHaveBeenCalledTimes(1);
  });
});
