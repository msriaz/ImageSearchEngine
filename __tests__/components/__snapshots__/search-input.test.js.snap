// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput Component should match the Search Input snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 8,
        "marginLeft": 8,
        "marginRight": 8,
        "marginTop": 8,
      },
    ]
  }
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Type something"
    style={
      Array [
        Object {
          "borderColor": "#E5E5E5",
          "borderStyle": "solid",
          "borderWidth": 1,
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
          "width": "NaNpx",
        },
      ]
    }
    value="cat"
  />
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "aliceblue",
          "justifyContent": "center",
        },
      ]
    }
  >
    <ActivityIndicator
      size="small"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "justifyContent": "center",
          },
        ]
      }
    />
  </View>
</View>
`;
